drivers/frequency/adf5355/adf5355.c

```diff
@@ -0,0 +1,518 @@
+				    ADF5355_REG2_FRAC2(dev->fract2);
+
+	if (dev->dev_id == ADF5356)
+		dev->regs[ADF5355_REG(13)] = ADF5356_REG13_MOD2_MSB(dev->mod2 >> 14) |
+					     ADF5356_REG13_FRAC2_MSB(dev->fract2 >> 14);
+
+	dev->regs[ADF5355_REG(6)] = ADF5355_REG6_OUTPUT_PWR(dev->outa_power) |
+				    ADF5355_REG6_RF_OUT_EN(dev->outa_en) |
+				    (dev->dev_id == ADF5355 ? ADF5355_REG6_RF_OUTB_EN(!dev->outb_en) :
+				     ADF4355_REG6_OUTPUTB_PWR(dev->outb_power) |
+				     ADF4355_REG6_RF_OUTB_EN(dev->outb_en)) |
+				    ADF5355_REG6_MUTE_TILL_LOCK_EN(dev->mute_till_lock_en) |
+				    ADF5355_REG6_CP_BLEED_CURR(cp_bleed) |
+				    ADF5355_REG6_RF_DIV_SEL(dev->rf_div_sel) |
+				    ADF5355_REG6_FEEDBACK_FUND(1) |
+				    ADF5355_REG6_NEG_BLEED_EN(cp_neg_bleed_en) |
+				    ADF5355_REG6_GATED_BLEED_EN(dev->cp_gated_bleed_en) |
+				    ADF5356_REG6_BLEED_POLARITY(dev->dev_id == ADF5356 ?
+						    dev->cp_bleed_current_polarity_en : 0) |
+				    ADF5355_REG6_DEFAULT;
+
+	dev->freq_req = freq;
+
+	return adf5355_reg_config(dev, dev->all_synced);
+}
+
+/**
+ * Get the output frequency of one channel.
+ * @param dev - The device structure.
+ * @param channel - The selected channel.
+ * @return The ouput frequency.
+ */
+static uint64_t adf5355_pll_fract_n_get_rate(struct adf5355_dev *dev,
+		uint32_t channel)
+{
+	uint64_t val, tmp;
+
+	val = (((uint64_t)dev->integer * ADF5355_MODULUS1) + dev->fract1) * dev->fpfd;
+	tmp = (uint64_t)dev->fract2 * dev->fpfd;
+	do_div(&tmp, dev->mod2);
+	val += tmp + ADF5355_MODULUS1 / 2;
+
+	do_div(&val, ADF5355_MODULUS1 *
+	       (1 << (channel == 1 ? 0 : dev->rf_div_sel)));
+
+	if (channel == 1)
+		val <<= 1;
+
+	return val;
+}
+
+/**
+ * Recalculate rate corresponding to a channel.
+ * @param dev - The device structure.
+ * @param chan - Channel number.
+ * @param rate - Channel rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf5355_clk_recalc_rate(struct adf5355_dev *dev, uint32_t chan,
+				uint64_t *rate)
+{
+	if (chan > dev->num_channels)
+		return FAILURE;
+
+	*rate = adf5355_pll_fract_n_get_rate(dev, chan);
+
+	return SUCCESS;
+}
+
+/**
+ * Set channel rate.
+ * @param dev - The device structure.
+ * @param chan - Channel number.
+ * @param rate - Channel rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf5355_clk_set_rate(struct adf5355_dev *dev, uint32_t chan,
+			     uint64_t rate)
+{
+	if (chan >= dev->num_channels)
+		return FAILURE;
+
+	return adf5355_set_freq(dev, rate, chan);
+}
+
+/**
+ * Calculate closest possible rate
+ * @param dev - The device structure
+ * @param rate - The desired rate.
+ * @param rounded_rate - The closest possible rate of desired rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf5355_clk_round_rate(struct adf5355_dev *dev, uint64_t rate,
+			       uint64_t *rounded_rate)
+{
+	*rounded_rate = rate;
+
+	return SUCCESS;
+}
```

Comment:
Is this necessary?

Reply:
I mentioned in the note of the PR:

> Note: `adf5355_clk_recalc_rate`, `adf5355_clk_set_rate` and `adf5355_clk_round_rate` function were added to align the driver with the linux version and ease the integration with certain projects.

See another example: 
https://github.com/analogdevicesinc/no-OS/blob/ca092f47e8ed60e7fb94b265fdcf27da12b6c3d9/drivers/frequency/adf4371/adf4371.c#L709-L721